lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee

```diff
@@ -61,25 +72,53 @@ define((require, exports, module) ->
       return stack[stack.length - 1]
 
     # specialized rules for context
-    basicRules = (tokenType) -> [
+    basicRules = (tokenType) ->
       if (tokenType?)
         addToken = "." + tokenType
       else
         addToken = ""
-      {
-        token: "comment"
-        regex: "%.*$"
-      }
+      return [
+        { token: "comment" + addToken, regex: "%.*$" }
+        { token: "lparen" + addToken, regex: "[[({]" }
+        { token: "rparen" + addToken, regex: "[\\])}]" }
+        { token: "storage.type" + addToken, regex: "\\\\[a-zA-Z]+" }
+        { token: "constant.character.escape" + addToken, regex: "\\\\[^a-zA-Z]?" }
+        { defaultToken : "text" + addToken }
+      ]
 
-      {
-        token: "lparen"
-        regex: "[[({]"
+    beginRule = (text = "\\w*", pushedState = "start") ->
+      return {
+        token: [
+          "storage.type"
+          "lparen"
+          "variable.parameter"
+          "rparen"
+        ]
+        regex: "(\\\\(?:begin))({)(" + text + ")(})"
+        next: pushState(pushedState)
       }
-      {
-        token: "rparen"
-        regex: "[\\])}]"
+
+    endRule = (text = "\\w*") ->
+      return {
+        token: [
+          "storage.type"
+          "lparen"
+          "variable.parameter"
+          "rparen"
+        ]
+        regex: "(\\\\(?:end))({)(" + text + ")(})"
+
+        next: popState
       }
 
+    specificTokenForState = {}
```

Comment:
I actually tried, and funny thing happened: `specificTokenForState` had strings `"LIST_STATE"`, `"EQUATION_STATE"` and so on as keys.